### PR TITLE
[Refactor] Test/Query mode refactor

### PR DESCRIPTION
**Pull request by mproch**
This change should not impact model API

- Move implementation details of result collection to interpreter
- Extract ResultCollector interface to replace inheritance with composition
- Move ResultCollector implementations to places closer to use
- Next steps: use ResultCollector to handle sinks, custom nodes etc. and remove e.g. SinkInvocationCollector

### CODE REVIEW

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/resultcollector/ResultCollector.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    @request - object that should be recorded (e.g. REST request)
+    @mockValue - value to return when invocation is stubbed, None means normal invocation should proceed
+    @action - e.g. REST service invocation, stubbed e.g. during tests. CollectableAction contains not only result, but also can contain
+      "raw" format, e.g. REST service response
+    @names - in more complex scenarios we may want to distinguish different invocations
+   */
+  def collectWithResponse[A, F[_] : Monad](contextId: ContextId,
+                                           nodeId: NodeId,
+                                           serviceRef: String,
+
```
Comment (lciolecki) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/resultcollector/ResultCollector.scala:27]: Remove new line?

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/nodecompilation/NodeDataValidator.scala [hunk 2]
```diff
@@ -35,7 +35,7 @@ object NodeDataValidator {
         case spel: SpelExpressionParser => spel.typingDictLabels
       }
       val compiler = new NodeCompiler(modelData.processWithObjectsDefinition,
```
Comment (lciolecki) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/nodecompilation/NodeDataValidator.scala:37]: Maybe better way is create companion object `NodeCompiler` with default `PreventInvocationCollector`?

**Review by lciolecki — APPROVED** [review]
LGTM

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/resultcollector/ResultCollector.scala (hunk 1 above)
Comment (mproch) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/resultcollector/ResultCollector.scala:27]: :+1:

**Pull request merged**